security/pkg/nodeagent/cache/secretcache.go

```diff
@@ -220,9 +209,31 @@ func (sc *SecretManagerClient) CallUpdateCallback(resourceName string) {
 	}
 }
 
-// GenerateSecret generates new secret and cache the secret, this function is called by SDS.StreamSecrets
-// and SDS.FetchSecret. Since credential passing from client may change, regenerate secret every time
-// instead of reading from cache.
+// getCachedSecret: retrieve cached Secret Item (workload-certificate/workload-root) from secretManager client
+func (sc *SecretManagerClient) getCachedSecret(resourceName string) (secret *security.SecretItem) {
+	var rootCertBundle []byte
+	if c := sc.cache.GetWorkload(); c != nil {
```

Comment:
I don't see why they shouldn't be coupled. Our design decision is to ALWAYS include the workload-root when responding to the ROOTCA request. If the workload cache is empty, this implies that there is a new certificate request that may result in a new root. Therefore, if this fails, we grab a lock and wait (the same way we do for certificate requests) until requests to the cache or no longer empty OR we fire a request to the CA asking for the new certificate chain.

Reply:
Our older code was buggy: It still fulfilled the ROOTCA request based on the root of the cert-chain obtained from the WORKLOAD request. However, the way it decoupled the two requests was as follows: the ROOTCA request would try to get the root, then retry + backoff a limited number of times, if that failed, they would return nil **INSTEAD** of waiting for the workload certificate request to come through which contained the root

Reply:
Yes, I think this is OK because the workload certificate always needs to be obtained. I can't think of any use case that you want the roots before the workload certificate is ready.